keras/models.py

```diff
@@ -73,10 +73,10 @@ def weighted(y_true, y_pred, weights, mask=None):
         weighted = filtered_weights * obj_output
         if mask is None:
             # Instead of calling mean() here, we divide by the sum of filtered_weights.
-            return weighted.sum() / filtered_weights.sum()
+            return weighted.sum() / (filtered_weights.sum() + 1e-20)
         else:
             filtered_mask = mask[weights.nonzero()[:-1]]
-            return weighted.sum() / (filtered_mask * filtered_weights).sum()
+            return weighted.sum() / ((filtered_mask * filtered_weights).sum() + 1e-20)
```

Comment:
please, see a few lines above, all the `sample_weight` values that are equal zero are filtered out before hand. Unless somebody changed that later.

Also, this is overlapping with what we are building for the objective decorator. We will give more control to the user about how he wants to use his sample weights. We wont be forcing anything. See #802